Adds HDF io Module

jlaura opened the pull request:
This is something to discuss tomorrow re: API.  Questions:

* Descriptors and keypoints are stored in an HDF5 file.  How does the user specify when these might exist already?  Is this done into the network?  Or when the adjacency matrix / file list are read?
* How do we let the user indicate that they want to save the keypoints?
* Image identifiers.  ISIS uses serial numbers.  How do we want to handle this?

Please do not merge.

jlaura commented:
Good to know - failure is due to a Cython dependency and this OS.  Something to watch for.

jlaura commented:
This is finally ready for review.  Anyone can have a whack at it whenever is convenient.

Something to discuss - right now the image name is used as the key in the HDF.  I really dislike this approach.  My preferred approach would be to compute an MD5 hash of the image.  This would work fine in the case of most formats, but not ISIS, because ISIS can modify the PVL header and append history.  Those actions change the hash.

An potential solution would be to compute the hash using the binary data if we hit an ISIS cube.  I do not want to have custom, `if isis` all over the place...

Kelvinrr commented on autocnet/graph/network.py:
```diff
@@ -239,7 +241,64 @@ def extract_features(self, method='orb', extractor_parameters={}):
+
+        # Cleaner way to do this?
+        if nodes:
+            for i, n in self.nodes_iter(nodes, data=True):
```
Is this valid? I can't find anywhere in the networkx docs that shows `nodes_iter` can traverse a subset of nodes.

Kelvinrr commented on autocnet/graph/network.py:
```diff
@@ -239,7 +241,64 @@ def extract_features(self, method='orb', extractor_parameters={}):
+            for i, n in self.nodes_iter(data=True):
+                n.save_features(hdf)
+
+        hdf = None
```
Why `hdf = none`? Wouldn't the `__del__` be called when `hdf` leaves the stack frame regardless and close the file?

Kelvinrr commented on autocnet/graph/network.py:
```diff
@@ -239,7 +241,63 @@ def extract_features(self, method='orb', extractor_parameters={}):
+            for i, n in self.nodes_iter(data=True):
+                n.load_features(hdf)
+
+        hdf = None
```
Probably should get rid of all of them for the sake of consistency.

jlaura commented:
Updated with an intentional `hdf=None`.  This ensures that the file is properly closed and does not need to wait for garbage collection.  Corrected the iterator in the load method.

kree commented on autocnet/fileio/io_hdf.py:
```diff
@@ -0,0 +1,84 @@
+            index = None
+        columns = clean_array.dtype.names
+
+        return pd.DataFrame(data=sarray, index=index, columns=columns)
```
PEP 8 wants a newline.

kree commented on autocnet/utils/tests/test_utils.py:
```diff
@@ -92,3 +92,12 @@ def test_make_homogeneous(self):
+        truth = np.array([(2.,'String'), (3.,"String2")],
+              dtype=[('bar', 'f4'), ('baz', 'S10')])
+        cleaned_array = utils.remove_field_name(starray, 'index')
+        np.testing.assert_array_equal(cleaned_array, truth)
```
PEP8 newline needed

kree commented on autocnet/utils/utils.py:
```diff
@@ -159,3 +159,28 @@ def make_homogeneous(points):
+    """
+    Given a numpy structured array, remove a column and return
+    a copy of the remaineder of the array
+
```
remaineder -> remainder

kree commented on autocnet/utils/utils.py:
```diff
@@ -159,3 +159,28 @@ def make_homogeneous(points):
+    if name in names:
+        names.remove(name)
+    b = a[names]
+    return b
```
PEP 8 newline needed.

kree commented:
@jlaura, @Kelvinrr asked me to take a quick look at this, so I added some quick PEP & documentation comments.
